scrapy_autoextract/providers.py

```diff
@@ -70,32 +147,50 @@ async def __call__(self):
                 max_query_error_retries=max_query_error_retries
             )
         except Exception:
-            self.stats.inc_value(f"autoextract/{page_type}/error/request")
+            self.stats.inc_value("autoextract/error/request")
             raise
 
+        self.stats.inc_value("autoextract/success")
+        self.response = response
+
+    async def _await_data(self):
+        response = None
+        while response is None:
+            response = self.response
+            await asyncio.sleep(0.2)
+
         data = response[0]
+        return data
 
-        if "error" in data:
-            self.stats.inc_value(f"autoextract/{page_type}/error/query")
-            raise QueryError(data["query"], data["error"])
 
-        self.stats.inc_value(f"autoextract/{page_type}/success")
-        return self.provided_class(data=data)
+class HTMLDataProvider(_Provider):
 
-    @classmethod
-    def register(cls):
-        """Register this provider for its provided class on scrapy-poet
-        registry. This will make it possible to declare provided class as
-        a callback dependency when writing Scrapy spiders.
-        """
-        register(cls, cls.provided_class)
+    provided_class = AutoExtractHTMLData
 
-    @classmethod
-    def get_page_type(cls) -> str:
-        """Page type is defined by the class attribute `item_key` available on
-        `autoextract_poet.page_inputs` classes.
+    def __before__(self):
+        self.extra = {
```

Comment:
why not in `__init__`?

Reply:
Because all providers are initialized despite being used or not.